engine/flink/process/src/main/scala/pl/touk/nussknacker/engine/process/typeinformation/internal/typedobject/TypedObjectBasedTypeInformation.scala

```diff
@@ -0,0 +1,193 @@
+
+  override def duplicate(): TypeSerializer[T] = duplicate(
+    serializers.map {
+      case (k, s) => (k, s.duplicate())
+    })
+  
+  override def copy(from: T): T = from
+
+  //???
+  override def copy(from: T, reuse: T): T = copy(from)
+
+  override def getLength: Int = -1
+
+  override def serialize(record: T, target: DataOutputView): Unit = {
+    serializers.foreach { case (key, serializer) =>
+      val valueToSerialize = get(record, key)
+      if (valueToSerialize == null) {
+        target.writeBoolean(true)
+      } else {
+        target.writeBoolean(false)
+        serializer.asInstanceOf[TypeSerializer[AnyRef]].serialize(valueToSerialize, target)
+      }
+    }
+  }
+
+  override def deserialize(source: DataInputView): T = {
+    //TODO: remove array allocation
+    val array: Array[AnyRef] = new Array[AnyRef](serializers.size)
+    serializers.indices.foreach { idx =>
+      array(idx) = if (!source.readBoolean()) {
+        serializersInstances(idx).asInstanceOf[TypeSerializer[AnyRef]].deserialize(source)
+      } else {
+        null
+      }
+    }
+    deserialize(array)
+  }
+
+  override def snapshotConfiguration(): TypeSerializerSnapshot[T] = snapshotConfiguration(serializers.map {
+    case (k, s) => (k, s.snapshotConfiguration())
+  })
+
+  override def deserialize(reuse: T, source: DataInputView): T = deserialize(source)
+
+  override def copy(source: DataInputView, target: DataOutputView): Unit = serializers.map(_._2).foreach(_.copy(source, target))
+
+  def snapshotConfiguration(snapshots: List[(String, TypeSerializerSnapshot[_])]): TypeSerializerSnapshot[T]
+
+  def deserialize(values: Array[AnyRef]): T
+
+  def get(value: T, name: String): AnyRef
+
+  def duplicate(serializers: List[(String, TypeSerializer[_])]): TypeSerializer[T]
+}
+
+abstract class TypedObjectBasedSerializerSnapshot[T] extends TypeSerializerSnapshot[T] with LazyLogging {
+
+  protected var serializersSnapshots: List[(String, TypeSerializerSnapshot[_])] = _
+
+  protected def nonEqualKeysCompatible: Boolean
+
+  def this(serializers: List[(String, TypeSerializerSnapshot[_])]) = {
+    this()
+    this.serializersSnapshots = serializers
+  }
+
+  override def getCurrentVersion: Int = 0
+
+  override def writeSnapshot(out: DataOutputView): Unit = {
+    out.writeInt(serializersSnapshots.size)
+    serializersSnapshots.foreach {
+      case (k, v) =>
+        out.writeUTF(k)
+        out.writeUTF(v.getClass.getCanonicalName)
+        v.writeSnapshot(out)
+    }
+  }
+
+  override def readSnapshot(readVersion: Int, in: DataInputView, userCodeClassLoader: ClassLoader): Unit = {
+    val size = in.readInt()
+    serializersSnapshots = (0 until size).map { _ =>
+      val key = in.readUTF()
+      val snapshot = InstantiationUtil.resolveClassByName(in, userCodeClassLoader)
+        .getConstructor().newInstance().asInstanceOf[TypeSerializerSnapshot[_]]
+      snapshot.readSnapshot(1, in, userCodeClassLoader)
```

Comment:
Is this `1` is correct? Shouldn't we use `TypeSerializerSnapshot.readVersionedSnapshot` instead? Many serializers currently uses other versions.